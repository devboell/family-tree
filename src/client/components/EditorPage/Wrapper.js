import styled from 'styled-components'
import { pageBlock } from 'globalStyles'


const Wrapper = styled.div`
  ${pageBlock}
  color: white;
  background-color: rgb(64,64,64);
`

export default Wrapper
