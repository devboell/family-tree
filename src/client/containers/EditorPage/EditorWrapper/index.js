import styled from 'styled-components'


const EditorWrapper = styled.div`
  display: flex;
  flex-flow: column nowrap;
  flex-basis: 80%;
  padding-left: 50px;
`

export default EditorWrapper
