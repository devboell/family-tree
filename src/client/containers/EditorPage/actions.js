import {
  SET_SELECTED_PERSON_ID,
  SET_CREATE_MODE,
  SET_IS_REMOVING,
} from './constants'

export const setSelectedPersonId = id => ({
  type: SET_SELECTED_PERSON_ID,
  id,
})

export const setCreateMode = flag => ({
  type: SET_CREATE_MODE,
  flag,
})

export const setIsRemoving = flag => ({
  type: SET_IS_REMOVING,
  flag,
})

export const clearSelection = () => ({
  type: SET_SELECTED_PERSON_ID,
  id: 'no_selection',
})
