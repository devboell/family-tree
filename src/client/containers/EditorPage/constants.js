export const editModes = {
  DISABLED: 'disabled',
  CREATE: 'create',
  UPDATE: 'update',
}


export const SET_EDIT_MODE = 'SET_EDIT_MODE'
export const SET_SELECTED_PERSON = 'SET_SELECTED_PERSON'
