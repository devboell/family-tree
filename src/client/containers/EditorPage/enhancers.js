import { gql, graphql, compose } from 'react-apollo'

const personInfoFragment = gql`
  fragment personInfo on Person {
    name
    title
    birthDate
    deathDate
    gender
    isMonarch
    bornToId
    partnerIds
    partners {
      relationId
      partner {
        id
        name
      }
      children {
        name
      }
    }
  }
`

export const PERSONS_QUERY = gql`
  query personsQuery {
    persons {
      id
      ...personInfo
    }
  }
  ${personInfoFragment}
`

export const RELATIONS_QUERY = gql`
  query relationsQuery {
    relations {
      id
      person1 {
        name
      }
      person2 {
        name
      }
    }
  }
`

export const CREATE_PERSON_MUTATION = gql`
  mutation createPersonMutation(
    $name: String
    $title: String
    $birthDate: Int
    $deathDate: Int
    $gender: String
    $isMonarch: Boolean
    $bornToId: ID
    $partners: [PartnerInput]
  ) {
    createPerson(
      name: $name
      title: $title
      birthDate: $birthDate
      deathDate: $deathDate
      gender: $gender
      isMonarch: $isMonarch
      bornToId: $bornToId
      partners: $partners
    ) {
      id
      ...personInfo
    }
  }
  ${personInfoFragment}
`

export const UPDATE_PERSON_MUTATION = gql`
  mutation updatePersonMutation(
    $id: ID!
    $name: String
    $title: String
    $birthDate: Int
    $deathDate: Int
    $gender: String
    $isMonarch: Boolean
    $bornToId: ID
    $partners: [PartnerInput]
  ) {
    updatePerson(
      id: $id
      name: $name
      title: $title
      birthDate: $birthDate
      deathDate: $deathDate
      gender: $gender
      isMonarch: $isMonarch
      bornToId: $bornToId
      partners: $partners
    ) {
      id
      ...personInfo
    }
  }
  ${personInfoFragment}
`
const REMOVE_PERSON_MUTATION = gql`
  mutation removePersonMutation(
    $id: ID!
  ) {
    removePerson(
      id: $id
    )
  }
`

export default compose(
  graphql(RELATIONS_QUERY, { name: 'relationsQuery' }),
  graphql(PERSONS_QUERY, { name: 'personsQuery' }),
  graphql(CREATE_PERSON_MUTATION, {
    props: ({ mutate }) => ({
      createPersonMutation: values => mutate({
        variables: values,
        refetchQueries: [
          {
            query: PERSONS_QUERY,
          },
          {
            query: RELATIONS_QUERY,
          },
        ],
      }),
    }),
  }),
  graphql(UPDATE_PERSON_MUTATION, {
    props: ({ mutate }) => ({
      updatePersonMutation: values => mutate({
        variables: values,
        refetchQueries: [
          {
            query: PERSONS_QUERY,
          },
          {
            query: RELATIONS_QUERY,
          },
        ],
      }),
    }),
  }),
  graphql(REMOVE_PERSON_MUTATION, {
    props: ({ mutate }) => ({
      removePersonMutation: values => mutate({
        variables: values,
        refetchQueries: [
          {
            query: PERSONS_QUERY,
          },
          {
            query: RELATIONS_QUERY,
          },
        ],
      }),
    }),
  }),
)
