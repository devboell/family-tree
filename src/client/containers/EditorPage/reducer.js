import {
  editModes,
  SET_EDIT_MODE,
  SET_SELECTED_PERSON } from './constants'

const initialState = {
  editMode: editModes.DISABLED,
  selectedPerson: {},
}

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_EDIT_MODE:
      return { ...state, editMode: action.mode }
    case SET_SELECTED_PERSON:
      return { ...state, selectedPerson: action.person }

    default: return state
  }
}
