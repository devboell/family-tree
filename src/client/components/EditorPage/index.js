import React from 'react'

import PersonList from 'containers/PersonList'
import Editor from 'containers/Editor'

import Wrapper from './Wrapper'

const EditorPage = () =>
  <Wrapper>
    <PersonList>
      <Editor />
    </PersonList>
  </Wrapper>

export default EditorPage
