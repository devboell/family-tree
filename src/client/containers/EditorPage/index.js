import React from 'react'
import { PropTypes } from 'prop-types'
import { connect } from 'react-redux'

import { selectPerson, createPerson, removePerson } from './thunks'
import enhancers from './enhancers'
import { editModes } from './constants'

import EditorControls from '../../components/EditorControls'
import PersonEditor from '../../components/PersonEditor'
import PersonList from '../../components/PersonList'
import Wrapper from './Wrapper'
import EditorWrapper from './EditorWrapper'

const EditorPage = ({
  relationsQuery,
  personsQuery,
  relationsQuery: { relations },
  personsQuery: { persons },
  createPersonMutation,
  updatePersonMutation,
  removePersonMutation,
  editMode,
  onSelectPerson,
  onCreatePerson,
  onRemovePerson,
  selectedPerson,
}) => {
  const convertPartnerValues = values => (
    {
      ...values,
      partners: values.partners.map(partner => (
        {
          relationId: partner.relationId,
          partnerId: partner.partner.id,
        }
      )),
    }
  )

  const onSubmit = (editMode === editModes.CREATE)
    ? (values) => {
      createPersonMutation(convertPartnerValues(values)).then((result) => {
        onSelectPerson(result.data.createPerson)
      })
    }
    : (values) => { updatePersonMutation(convertPartnerValues(values)) }

  const editorProps = {
    disabled: editMode === editModes.DISABLED,
    onSubmit,
    initialValues: selectedPerson,
    partners: selectedPerson.partners,
    persons,
    relations,
  }

  return (
    relationsQuery.loading || personsQuery.loading
    ? <div>loading ...</div>
    : <Wrapper>
      <PersonList {...{ persons, onSelectPerson, selectedPerson }} />
      <EditorWrapper>
        <EditorControls
          {...{
            editMode,
            selectedPerson,
            onCreatePerson,
            onRemovePerson,
            removePersonMutation,
          }}
        />
        <PersonEditor {...editorProps} />
      </EditorWrapper>
    </Wrapper>
  )
}
EditorPage.propTypes = {
  relationsQuery: PropTypes.object.isRequired,
  personsQuery: PropTypes.object.isRequired,
  createPersonMutation: PropTypes.func.isRequired,
  updatePersonMutation: PropTypes.func.isRequired,
  removePersonMutation: PropTypes.func.isRequired,
  editMode: PropTypes.string.isRequired,
  onSelectPerson: PropTypes.func.isRequired,
  onCreatePerson: PropTypes.func.isRequired,
  onRemovePerson: PropTypes.func.isRequired,
  selectedPerson: PropTypes.object.isRequired,
}

const mapStateTopProps = state => ({
  editMode: state.editor.editMode,
  selectedPerson: state.editor.selectedPerson,
})

const mapDispatchToProps = dispatch => ({
  onSelectPerson: id => dispatch(selectPerson(id)),
  onCreatePerson: () => dispatch(createPerson()),
  onRemovePerson: () => dispatch(removePerson()),
})

export default enhancers(connect(mapStateTopProps, mapDispatchToProps)(EditorPage))
