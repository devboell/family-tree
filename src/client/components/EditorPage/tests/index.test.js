import React from 'react'
import EditorPage from '../index'

it('EditorPage component, it renders without crashing', () => {
  const wrapper = shallow(<EditorPage />)
  expect(wrapper).toMatchSnapshot()
})
