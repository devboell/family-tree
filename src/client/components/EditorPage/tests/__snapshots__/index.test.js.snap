// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorPage component, it renders without crashing 1`] = `
<styled.div>
  <Connect(Apollo(PersonList))>
    <Connect(Apollo(Apollo(Apollo(Apollo(Editor))))) />
  </Connect(Apollo(PersonList))>
</styled.div>
`;
